### PR TITLE
Update Python API to support REST API v2

### PR DESCRIPTION
**Pull request by edewata**
The `PKIClient` class has been added to replace `PKIConnection` as the main access point to PKI services. By default it will use REST API v2, then fall back to v1 if it's not available. Optionally, `PKIClient` can be configured to use a specific REST API version.

The `InfoClient`, `CertClient`, `AccountClient`, and `UserClient` classes have been added/updated to construct the proper REST URL according to the REST API version in `PKIClient`.

The `pki-healthcheck` has been updated to use `PKIClient`. Some simple Python scripts have also been added to demonstrate how to use `PKIClient`.

New tests have been added to run these scripts against the current CA and KRA which support both REST API v1 and v2 and also against an older CA that only supports REST API v1.

### CODE REVIEW

**Comment by rcritten** [comment]
It is unexpected that one has to always pass in port when instantiating PKIClient when using a URL. If no port is provided for an http/https connection the port should default to 80/443.

**Comment by rcritten** [comment]
list_certs is not working:
```
import pki.ca
import pki.cert
import pki.client
import pki.systemcert

pki_client = pki.client.PKIClient(url='https://ipa.example.test:443', verify=True)
print("API version %s" % pki_client.get_api_version())

ca_client = pki.ca.CAClient(pki_client)
cert_client = pki.cert.CertClient(ca_client)

print(cert_client.get_cert(1))

print(cert_client.list_certs(size=1))
```
Fails with a huge traceback but the bottom-line error is:

*** requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://ipa.example.test:443/ca/v2/certs/search?size=1

**Comment by edewata** [comment]
> It is unexpected that one has to always pass in port when instantiating PKIClient when using a URL. If no port is provided for an http/https connection the port should default to 80/443.

I've updated the PR to fix this.

**Comment by edewata** [comment]
> pki_client = pki.client.PKIClient(url='https://ipa.example.test:443', verify=True)
> *** requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://ipa.example.test:443/ca/v2/certs/search?size=1

I think this is because IPA currently doesn't provide a proxy for REST API v2:
https://github.com/freeipa/freeipa/blob/master/install/share/ipa-pki-proxy.conf.template#L29-L35

**Comment by rcritten** [comment]
> I think this is because IPA currently doesn't provide a proxy for REST API v2: https://github.com/freeipa/freeipa/blob/master/install/share/ipa-pki-proxy.conf.template#L29-L35

I was using the python classes directly and not going through IPA.

**Comment by edewata** [comment]
> > I think this is because IPA currently doesn't provide a proxy for REST API v2: https://github.com/freeipa/freeipa/blob/master/install/share/ipa-pki-proxy.conf.template#L29-L35
> 
> I was using the python classes directly and not going through IPA.

Yes, but port 443 is owned by IPA and with the current configuration it's not forwarding the request to PKI.

**Comment by edewata** [comment]
@fmarco76 @rcritten Please take a look at the updated PR. I think all CI failures are fixed now (except for the known issues), and this change should not affect IPA since IPA is not using it yet. Once this is merged we can update other classes & methods to support REST API v2. Thanks!

**Review by fmarco76 — APPROVED** [review]
I have several concerns but the overall looks good.Feel free to update/merge this so it is possible to continue  integrating the other APIs and improve the design.

File: base/common/python/pki/client.py
Comment (edewata) [reply base/common/python/pki/client.py:440]: Hmm.. do you have some examples? Ideally the Python API is supposed to take params that can be used for both REST API v1 and v2, then call either REST API v1 or v2 depending on the availability and client's requirements, then parse the server's response and put it into objects that can be used by the Python client regardless of which REST API version was actually used.

File: base/common/python/pki/ca.py
Comment (edewata) [reply base/common/python/pki/ca.py:14]: The updated Python API is more consistent with the Java API, so the `PKIClient` serves as the main entry point for the client to set up the connection and the authentication, then the operations are grouped by subsystems (e.g. CA, KRA). Ideally the subsystem classes should inherit from a `SubsystemClient` class just like in Java API, but that can be added later. Having said that, I'm open to client API redesign as long as it's consistently done in Python and Java API.

**Comment by edewata** [comment]
@fmarco76 Thanks! I've clean up the PR as you suggested and will merge after the CI is done. Feel free to continue the client API design discussion.

**Comment by fmarco76** [comment]
> CI is done

Python CI is failing for a deprecated library. We should schedule the work to fix this since we are working on the python code

**Comment by edewata** [comment]
> Python CI is failing for a deprecated library. We should schedule the work to fix this since we are working on the python code

Looks like the container tests are failing too, but it doesn't seem to be related to this PR. I'll take a look after this.

**Pull request merged**
